src/live/embed.rs

```diff
@@ -503,14 +504,41 @@ struct InitrdEmbedArea {
     initrd: Initrd,
 }
 
+#[derive(Deserialize)]
+struct IgnInfo {
+    file: String,
+    offset: Option<u64>,
+    length: Option<usize>,
+}
+
 impl InitrdEmbedArea {
     pub fn for_iso(iso: &mut IsoFs) -> Result<Self> {
+        let igninfo: IgnInfo = match iso.get_path(COREOS_IGNINFO_PATH) {
+            Ok(record) => {
+                let f = record.try_into_file()?;
+                serde_json::from_reader(iso.read_file(&f).context("reading igninfo")?)
+                    .context("decoding igninfo")?
+            }
+            // old ISO without info JSON; assume ignition.img
+            Err(e) if e.is::<iso9660::NotFound>() => IgnInfo {
+                file: COREOS_INITRD_DEFAULT_EMBED_PATH.to_string(),
+                offset: None,
+                length: None,
+            },
+            Err(e) => return Err(e),
+        };
+
         let f = iso
-            .get_path(COREOS_INITRD_EMBED_PATH)
+            .get_path(&igninfo.file.to_uppercase())
             .context("finding initrd embed area")?
             .try_into_file()?;
+        let file_offset = igninfo.offset.unwrap_or(0);
+        let iso_offset = f.address.as_offset() + file_offset;
+        let length = igninfo
+            .length
+            .unwrap_or(f.length as usize - file_offset as usize);
```

Comment:
The double `as usize` casting isn't great here. One is from `u32` which seems reasonable but the other is from `u64` which seems more questionable. I brushed over this since in practice we know it'll fit with the offset values we're expecting. If we want to be more careful though, we could do this subtraction in u64 mode and then `TryFrom` the result into a `usize`.

Reply:
A valid offset can't be larger than the 32-bit maximum ISO9660 file size, and an invalid one can underflow the subtraction (in release builds) even with `u64`.  The consequence would be a short read error in `Region::read()`, which doesn't seem too bad.  I'd say we can leave it as is.